src/ax/marshaling/json.go

```diff
@@ -0,0 +1,108 @@
+package marshaling
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"mime"
+	"reflect"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+)
+
+const (
+	// JSONContentType is the content-type name for JSON-encoded messages.
+	JSONContentType = "application/json"
+)
+
+// MarshalJSON marshals a ProtocolBuffers messages to JSON
+// representation and returns a MIME content-type that identifies the particular
+// message protocol.
+func MarshalJSON(msg proto.Message) (ct string, data []byte, err error) {
+
+	n := proto.MessageName(msg)
+	if n == "" {
+		err = fmt.Errorf(
+			"can not marshal '%s', protocol is not registered",
+			reflect.TypeOf(msg),
+		)
+		return
+	}
+
+	b := new(bytes.Buffer)
+
+	m := jsonpb.Marshaler{
+		EmitDefaults: false,
+		EnumsAsInts:  false,
+		Indent:       "  ",
+		OrigName:     false,
+	}
+
+	ct = mime.FormatMediaType(
+		JSONContentType,
+		map[string]string{"proto": n},
+	)
+
+	err = m.Marshal(b, msg)
+	data = b.Bytes()
+
+	return
+}
+
+// UnmarshalJSON decodes JSON-encoded data into `proto.Message`
+func UnmarshalJSON(ct string, data []byte) (proto.Message, error) {
+
+	ctn, p, err := mime.ParseMediaType(ct)
+	if err != nil {
+		return nil, err
+	}
+
+	return UnmarshalJSONParams(ctn, p, data)
+
+}
+
+// UnmarshalJSONParams unmarshals a JSON-encoded messages using
```

Comment:
`messages` should be `message`, this was probably just copied from the PB docs where I made this typo